### DISCUSSION TITLE
waitFor devcontainer.json key not working, output of lifecycle scripts suppressed

### DISCUSSION THREAD

**Discussion by jnewland** [community/community#9644] [Codespaces]
https://github.com/jnewland/devcontainer-example/tree/waitFor contains a devcontainer configuration that demonstrates that the `waitFor` [lifecycle script configuration key](https://code.visualstudio.com/docs/remote/devcontainerjson-reference#_lifecycle-scripts) does not behave as documented on Codespaces.

As configured, I expect that all three of `onCreateCommand`, `updateContentCommand`, and `postCreateCommand` will be completed before I am able to connect to the Codespace, interact with files, open Terminals, and run commands. Instead, I'm connected to the Codespace while `onCreateCommand` is still running, as is evident by opening the Codespace creation log immediately. This has created confusion for users of a Codespaces deployment I'm working as it's not immediately evident when all setup tasks have completed.

#### Codespaces behavior: IDE is interactive while lifecycle scripts still in progress, output visible only only request

<img width="1136" alt="image" src="https://user-images.githubusercontent.com/47/148121217-9a33ebaa-b3d6-4d18-bdcc-4744c4ed4030.png">

When opening the same repository using VS Code Remote Containers locally, I'm presented with a "Configuring..." Terminal that displays the progress of these commands before connecting and allowing interactive use of the IDE. I would greatly prefer that Codespaces behaved similarly to ensure that users are only able to interact with the IDE once it's fully functional!

#### VS Code Remote Containers behavior: full usage of IDE blocked until waitFor, output visible by default

<img width="1136" alt="image" src="https://user-images.githubusercontent.com/47/148120998-c7b90e8e-50c5-4d7e-8494-734c42e87410.png">

Additionally, VS Code Remote Containers presents the *output* from lifecycle scripts to users in this "Configuring..." Terminal, which I've leveraged in a handful of repositories to provide usage instructions to contributors. I would greatly prefer that Codespaces allowed me to present the same usage instructions to users in the same way: by echoing them to stdout in a Terminal that is presented to the user by default.

**Comment by jnewland** [comment]
Thanks for documenting this lack of compatibility recently! Do you have any suggested workarounds for displaying a set of usage instructions to users in a way that is consistent between VS Code Remote Containers and Codespaces?

**Comment by sayhiben** [comment]
The "Welcome to Codespaces" message is just a text file; Codespaces checks for its existence in multiple directories, so if one of your setup scripts copies your own custom text file to one of those paths, you can customize this message. For example, I've placed this into my `onCreateCommand`:

```shell
echo "Copy first-run-notice.txt to Codespace"
sudo cp --force .devcontainer/first-run-notice.txt /usr/local/etc/vscode-dev-containers/first-run-notice.txt
```

**Comment by jnewland** [comment]
Thanks for the suggestion, @sayhiben! Is it possible to display a usage message without requiring the user to open a terminal like you can by echoing output in one of the setup scripts when using VS Code Remote Containers? I'm specifically trying to guide folks that aren't experienced users of VS Code towards using some built in features instead of opening a terminal as they might be used to.

> The "Welcome to Codespaces" message is just a text file

Additionally, is this a feature of VS Code or a feature of some of the devcontainer base images or scripts? I notice that in a repository that uses its own Debian-based devcontainer base image that no `/usr/local/etc/vscode-dev-containers/` exists, and found https://github.com/microsoft/vscode-dev-containers/blob/58bf3c5921f25c6ebb97f6045df38a93a6470b3d/script-library/common-debian.sh#L238-L248 leading me to believe that this functionality is expected to be implemented by base images rather than handled at the editor level.

**Comment by JustinGrote** [comment]
We do this as a workaround in our devcontainer dockerfile to do setup tasks prior to the UI loading  (this is using PowerShell but similar instructions work fine for bash/zsh) until `waitFor` is supported in codespaces

```docker
SHELL ["/usr/bin/pwsh", "-noprofile", "-command"]
USER vscode
COPY Scripts/Install-OhMyPosh.ps1 /tmp
RUN /tmp/Install-OhMyPosh.ps1
```
